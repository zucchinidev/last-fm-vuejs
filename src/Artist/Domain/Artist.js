export class Artist {
  constructor ({ name, url, associateImages: images } = {}) {
    this.name = name
    this.url = url
    this.images = images
  }
}
