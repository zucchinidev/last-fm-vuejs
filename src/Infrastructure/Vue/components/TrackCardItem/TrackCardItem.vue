<template>
  <v-flex v-bind="breakPoints" wrap v-if="track && track.album">
    <v-card>
      <v-container fluid grid-list-lg>
        <v-layout row>
          <v-flex xs7>
            <div class="text-xs-center">
              <template v-if="!extendedData">
                <div class="title">
                  <p>Album</p>
                  <p class="album-name">
                    {{track.album.name}}
                  </p>
                </div>
                <p class="track-name">
                  {{track.name}}
                </p>
                <p>Duration
                  <v-icon>arrow_right_alt</v-icon>
                  {{track.duration | ms-to-mm}}
                </p>
              </template>
              <template v-if="extendedData">
                <p class="title">Album
                  <v-icon>arrow_right_alt</v-icon>
                  {{track.album.name}}
                </p>
                <p>Track name
                  <v-icon>arrow_right_alt</v-icon>
                  {{track.name}}
                </p>
                <p>Release
                  <v-icon>arrow_right_alt</v-icon>
                  {{track.album.release}}
                </p>
                <p> Popularity
                  <v-icon>arrow_right_alt</v-icon>
                  {{track.popularity}}
                </p>
                <div>
                  <p class="display-1 text-xs-left">
                    Artists
                    <v-icon>trending_down</v-icon>
                  </p>
                  <ul>
                    <li
                      class="text-xs-center"
                      :key="index"
                      v-for="(artist, index) in track.album.artists">
                      <v-chip>
                        {{artist}}
                      </v-chip>
                    </li>
                  </ul>
                </div>
              </template>
            </div>
          </v-flex>
          <v-flex xs5>
            <v-card-media
              :src="imageSrc"
              height="125px"
              contain
            ></v-card-media>
            <v-card-actions v-if="!extendedData">
              <v-spacer></v-spacer>
              <v-btn class="play-arrow" icon @click="selectTrack">
                <v-icon>play_arrow</v-icon>
              </v-btn>
              <v-btn icon @click="goToTrack">
                <v-icon>remove_red_eye</v-icon>
              </v-btn>
              <v-spacer></v-spacer>
            </v-card-actions>
          </v-flex>
        </v-layout>
      </v-container>
    </v-card>
  </v-flex>
</template>
<script>
export default {
  name: 'TrackCardItem',
  props: {
    /** @type Track */
    track: {
      type: Object,
      required: true
    },
    breakPointsWrapper: {
      type: Object,
      default: () => ({})
    },
    extendedData: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    imageSrc () {
      let url = ''
      if (Object.keys(this.track).length) {
        const image = this.extendedData ? this.track.getBigImage() : this.track.getSmallImage()
        if (image) {
          url = image.url
        }
      }
      return url
    },
    breakPoints () {
      const defaultBreakPoints = {
        'xs12': true,
        'sm6': true,
        'md4': true,
        'lg4': true
      }
      return {
        ...defaultBreakPoints,
        ...this.breakPointsWrapper
      }
    }
  },
  methods: {
    selectTrack () {
      this.$emit('selectTrack', this.track)
    },
    goToTrack () {
      this.$router.push({ name: 'TrackDetail', params: { id: this.track.id } })
    }
  }
}
</script>
<style lang="stylus">
  ul
    display flex
    list-style none
    margin-bottom 1rem
</style>
