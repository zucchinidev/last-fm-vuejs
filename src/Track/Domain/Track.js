const availableSizes = {
  small: 64,
  medium: 300,
  big: 640
}

export class Track {
  constructor ({ album, duration, name, popularity, images, id, previewUrl } = {}) {
    this.id = id
    this.album = album
    this.duration = duration
    this.name = name
    this.popularity = popularity
    this.images = images
    this.previewUrl = previewUrl
  }

  getSmallImage () {
    return this.images.find(image => image.width === availableSizes.small)
  }

  getBigImage () {
    return this.images.find(image => image.width === availableSizes.big)
  }

  getPreviewUrl () {
    return this.previewUrl
  }
}
