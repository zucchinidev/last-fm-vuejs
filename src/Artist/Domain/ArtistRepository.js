import { HttpRepository } from '../../Infrastructure/http/HttpRepository'
import config from '../../Infrastructure/lastFmApi/config'
import { Artist } from './Artist'
const urls = {
  getArtistsByCountry: ''
}

const configuration = {
  slug: 'contacts',
  baseURL: 'http://ws.audioscrobbler.com/2.0/'
}
export const ArtistRepository = {
  async getTopArtistsByCountry (country = 'spain') {
    const params = {
      method: 'geo.gettopartists',
      country,
      api_key: config.apiKey,
      format: 'json'
    }
    const { data } = await HttpRepository.get(urls.getArtistsByCountry, params, configuration)
    return data.topartists.artist.map(artist => new Artist({
      images: artist.image.map(image => ({url: image['#text'], size: image.size})),
      url: artist.url,
      name: artist.name
    }))
  }
}
