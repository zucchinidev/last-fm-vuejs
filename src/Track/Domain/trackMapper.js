import { Track } from './Track'

export default (data) => {
  const artists = (data.album.artists || []).map(artist => artist.name)
  const images = data.album.images || {}
  return new Track({
    album: {
      artists,
      name: data.album.name,
      release: data.album.release_date
    },
    images,
    duration: data.duration_ms,
    name: data.name,
    popularity: data.popularity
  })
}
