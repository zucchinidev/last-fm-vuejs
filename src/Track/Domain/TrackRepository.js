import { HttpRepository } from '../../Infrastructure/http/HttpRepository'
import trackMapper from './trackMapper'

const urls = {
  search: 'search'
}

const configuration = {
  baseURL: process.env.API_MUSIC_BASE_URL
}
export const TrackRepository = {
  async searchTrack (searchTerm) {
    const params = {
      q: searchTerm,
      type: 'track'
    }
    const { data } = await HttpRepository.get(urls.search, params, configuration)
    return data.tracks.items.map(trackMapper)
  }
}
