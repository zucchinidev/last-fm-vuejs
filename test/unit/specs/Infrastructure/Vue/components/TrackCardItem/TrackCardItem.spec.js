import TrackCardItem from '@Vue/components/TrackCardItem/TrackCardItem'
import { WrapComponent } from '../../../../../helpers'
import { Track } from '../../../../../../../src/Track/Domain/Track'
import * as fixture from '../../../../../fixtures/trackFixture'
import TrackCardItemPageObject from './TrackCardItemPageObject'

const { track: trackFixture } = fixture
describe('TrackCardItem.vue', () => {
  let wrapper
  /** @type Track */
  let track
  /** @type TrackCardItemPageObject */
  let page
  beforeEach(() => {
    track = new Track(trackFixture)
    wrapper = WrapComponent(TrackCardItem).withProps({track}).mount()
    page = new TrackCardItemPageObject(wrapper)
  })
  it('should render correct contents', async () => {
    await page.wait()
    expect(page.getAlbumTitle()).toEqual(trackFixture.album.name)
    expect(page.getTrackTitle()).toEqual(trackFixture.name)
    expect(page.getPoster()).toEqual(track.getSmallImage().url)
    page.clickPlayButton()
    expect(page.isSelectTrackEventEmitted()).toBeTruthy()
    expect(page.isSelectTrackEventEmitted()).toEqual([[track.id]])
  })

  it('must receive a track as property', () => {
    expect(wrapper.props().track).toEqual(track)
  })

  it('should has correct html', function () {
    page.matchSnapshot()
  })
})
