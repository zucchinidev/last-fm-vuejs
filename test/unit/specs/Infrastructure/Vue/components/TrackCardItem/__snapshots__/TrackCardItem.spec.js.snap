// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackCardItem.vue should has correct html 1`] = `
<div class="flex wrap xs12 sm6 md4 lg4">
    <div class="card" style="height: auto;">
        <div class="container fluid grid-list-lg">
            <div class="layout row">
                <div class="flex xs7">
                    <div class="text-xs-center">
                        <div class="title">
                            <p>Album</p>
                            <p class="album-name">
                                Esa Muchacha
                            </p>
                        </div>
                        <p class="track-name">
                            Esa Muchacha
                        </p>
                        <p>Duration
                            <i aria-hidden="true" class="icon material-icons">arrow_right_alt</i>
                            03:00
                        </p>
                        <!---->
                    </div>
                </div>
                <div class="flex xs5">
                    <div class="card__media" style="height: 125px;">
                        <div class="card__media__background"></div>
                        <div class="card__media__content"></div>
                    </div>
                    <div class="card__actions">
                        <div class="spacer"></div>
                        <button type="button" class="play-arrow btn btn--icon">
                            <div class="btn__content">
                                <i aria-hidden="true" class="icon material-icons">play_arrow</i>
                            </div>
                        </button>
                        <button type="button" class="btn btn--icon">
                            <div class="btn__content">
                                <i aria-hidden="true" class="icon material-icons">remove_red_eye</i>
                            </div>
                        </button>
                        <div class="spacer"></div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
`;
